### ISSUE TITLE
Easily Access Nested Data with DataStore

### ISSUE THREAD

**Issue by Kilo-Loco** [aws-amplify/amplify-flutter#642]
TL;DR - I want to access nested data easily (`post.author.name`)

I'm running into the issue of loading the nested/associated data (i.e. Post Author) with the current Flutter DataStore implementation. I believe that most devs would expect to get associated data back with a single query as opposed to multiple queries to fill in the missing data of in individual object in the list of objects. 

Building something like a social media post feed, you would need to query the list of Posts, and for each Post, the Author would have to be queried for each Post. So then Amplify is forcing the dev into the position of figuring out when to make that second query for each post: 

1. Immediately after the query for the Posts, which could then lead to unnecessary queries for Authors if those Posts are never shown on screen.
2. Just before the next Post is displayed on the screen, which makes for a complex setup for managing the data.

Regardless of the approach, the data should be cached into memory for faster lookups, and this further complicates a developer's implementation of a social media post feed. This process would only get more complicated if there are multiple `hasOne` relationships on Post.

However, if the implementation was to follow the Amplify iOS approach of lazy loading the data, a lot of these issues are resolved for the developer and make Amplify much easier to use. The ideal solution would be to access two levels of data using a single query (i.e `post.author.name`) and three levels should require the additional query, since that data isn't directly relevant to the initial query (i.e `post.author.company.name`). Having the ability to query three levels deep would be nice, but two levels is a must.

Example Schema:
```graphql
type Post @model @auth(rules: [{allow: public}]) {
  id: ID!
  body: String!
  author: User @connection
}

type User @model @auth(rules: [{allow: public}]) {
  id: ID!
  name: String!
}
```

**Comment by dnys1** [comment]
Hi @Kilo-Loco - apologies for your frustration and thank you for taking the time to create this issue and provide some feedback. You're absolutely right on all accounts and the Amplify team is working hard to fix this in coming versions. It's looking like we will be moving towards an implementation similar to what you've described which piggybacks off of iOS's lazy loading of nested models.

Please follow our tracking ticket for this issue and stay tuned in the coming releases! https://github.com/aws-amplify/amplify-flutter/issues/260

**Comment by HuiSF** [comment]
With amplify-flutter 0.2.2 release, nested models (root model `belongsTo` nested models) will be populated with data to provide easy access.

**Comment by OLTApp** [comment]
Thanks for the request.  Me too stuck with out this featured for nested data load. Any estimated timeline on when this feature will be released? Trying to decide to use GraphQL pausing instead or wait to use this datastore...

**Comment by HuiSF** [comment]
Hi @OLTApp at present, DataStore plugin support direct access to nested model when the root model `belongsTo` the nested model. E.g. a `Post` belongs to a `Blog`, when you access `post.blog` you can get the data of `Blog`.

For other relationships, currently requires separate queries. E.g. a `Post` has an `Author`:

```dart
final post = (await Amplify.DataStore.query(Post.classType, where: Post.ID.equal('post-id')))[0];
final postAuthor = (await Amplify.DataStore.query(Author.classType, where: Author.ID.equal(post.authorId)))[0];
```

**Comment by OLTApp** [comment]
> Hi @OLTApp at present, DataStore plugin support direct access to nested model when the root model `belongsTo` the nested model. E.g. a `Post` belongs to a `Blog`, when you access `post.blog` you can get the data of `Blog`.
> 
> For other relationships, currently requires separate queries. E.g. a `Post` has an `Author`:
> 
> ```dart
> final post = (await Amplify.DataStore.query(Post.classType, where: Post.ID.equal('post-id')))[0];
> final postAuthor = (await Amplify.DataStore.query(Author.classType, where: Author.ID.equal(post.authorId)))[0];
> ```

Thank you for the response. That is working. That is so crucial for my use case. Really appreciate it.

**Comment by offlineprogrammer** [comment]
Hey @OLTApp 

Thanks for your update...
I am closing this issue for now as per your confirmation above 

Regards
Mo

**Comment by jeremyrajan** [comment]
@offlineprogrammer  will this work if I am not using the datastore? `ModelQueries.list(Appointment.classType);`

**Comment by SilverLinings89** [comment]
Same question here

**Comment by deborshisaha** [comment]
Do we have a similar support for React then? I agree that DataStore should have 2 level query by default.
